Add strict type declarations and modernize codebase

BastianLedererIcinga opened the pull request:
Add strict type declarations to all method parameters, return types, and
properties throughout the codebase. Replace return type `self` with              
`static` for better fluent interface support. Remove redundant runtime                                                                                                                                                                                                                                                                                                          
type checks now enforced by the type system. Replace `list()` with array                                                                                                                                                                                                                                                                                                        
destructuring.                                                                                                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                                                  
Raise the minimum required `ipl/sql` and `ipl/stdlib` versions to adopt                                                                                                                                                                                                                                                                                                         
their strict type declarations.

requires Icinga/ipl-stdlib#63
requires Icinga/ipl-sql#96

BastianLedererIcinga committed:
Add non-breaking strict type declarations

Add strict type declarations to properties, function/method signatures, where
types are unambiguous and no inheritance  or interface contract is affected.

BastianLedererIcinga committed:
Replace return type `self` with `static`

sukhwinder33445 commented on src/Relation/BelongsToMany.php:
```diff
@@ -87,7 +87,7 @@ public function setThroughAlias(string $throughAlias): self
      * @return Model|Junction
      */
-    public function getThrough(): Model
+    public function getThrough(): Model|Junction
```
`Junction` is just a subclass of `Model`, please remove it.

sukhwinder33445 commented on src/ResolvedExpression.php:
```diff
@@ -3,14 +3,15 @@
 namespace ipl\Orm;
 
 use Generator;
+use gipfl\Protocol\Generic\AbstractStreamWrapper;
```
Please remove.

sukhwinder33445 commented on src/Query.php:
```diff
@@ -204,7 +200,7 @@ public function setFilter(Filter\Chain $filter)
      * @return $this
      */
-    public function disableDefaultSort($disable = true)
+    public function disableDefaultSort($disable = true): static
```
Add type bool please.

lippserd commented on src/Query.php:
```diff
@@ -704,7 +700,7 @@ public function peekAhead($peekAhead = true)
      *
      * @return \Generator
```
You can simplify this to `Generator`.

lippserd commented on src/Relation.php:
```diff
@@ -327,7 +313,7 @@ public function determineKeys(Model $source)
      * @return \Generator
      */
-    public function resolve()
+    public function resolve(): \Generator
```
Add `use` statement and simplify both PHPDoc and return type to `Generator`.

lippserd commented on src/Relation/BelongsToMany.php:
```diff
@@ -164,14 +162,14 @@ public function getTargetCandidateKey()
     }
 
-    public function resolve()
+    public function resolve(): \Generator
```
Add `use` statement.

lippserd commented on src/Exception/InvalidRelationException.php:
```diff
@@ -46,7 +46,7 @@ public function getRelation()
      * @return Model
      */
-    public function getModel()
+    public function getModel(): Model
```
This method may return `null`.

sukhwinder33445 commented on src/Common/SortUtil.php:
Please use `[...]` instead of `list()`

sukhwinder33445 commented on src/Compat/FilterProcessor.php:
Same here.

sukhwinder33445 commented on src/Behaviors.php:
```diff
@@ -16,32 +16,34 @@
     /**
      * Add a behavior
      *
      * @param PersistBehavior|PropertyBehavior|RetrieveBehavior|RewriteFilterBehavior $behavior
```
This is missing `RewriteColumnBehavior|RewritePathBehavior`.
I would suggest simplifying it by using `Behavior` instead.

sukhwinder33445 commented on src/Exception/InvalidColumnException.php:
Why did you not define the type for the first parameter?

sukhwinder33445 reviewed (commented):
1. I think we can define return types for `PersistBehavior|RetrieveBehavior` interfaces, as they are not overridden by any other module.
2. Please use `[...]` instead of `list()`
3. Please import the class instead of using qualifier in `PropertiesWithDefaults (line 10-11)`.
4. Please remove the unused  variable `$err` in `MillisecondTimestamp`.

Note for myself:
The following classes and interfaces have not yet been modernized, as they are overridden by some modules (TBD):
Classes:
- PropertyBehavior
- Binary|BoolCast|MillisecondTimestamp
- Model
- UnionModel
- ResultSet
- Relation (partially)
- Query (Partially)

Interfaces:
- QueryAwareBehavior
- RewriteColumnBehavior
- RewriteFilterBehavior

BastianLedererIcinga committed:
Use array destructuring instead of `list()`

sukhwinder33445 commented on src/Exception/InvalidRelationException.php:
There are only two use cases in ipl-orm and none outside of it.

sukhwinder33445 reviewed (commented):
Looks good. 
Please use backticks in the commits to highlight function, class, interface, etc. names.

sukhwinder33445 reviewed (dismissed):
Looks good to me.
@lippserd, please take a look.

lippserd committed:
Require newer `ipl/sql` and `ipl/stdlib` releases

Raise the minimum `ipl/sql` and `ipl/stdlib` requirements to versions that
include the strict-typing updates and PHP 8.5 support.

Drop the `autoload-dev` path hack to make `ipl\Tests\Sql\` available. Now that
`TestCase` is part of `ipl\Sql\Test`, it is available through the normal
autoloader.
